beacon-chain/rpc/beaconv1/blocks.go

```diff
@@ -24,20 +116,217 @@ func (bs *Server) ListBlockHeaders(ctx context.Context, req *ethpb.BlockHeadersR
+			Data: &blockfeed.ReceivedBlockData{SignedBlock: v1alpha1Block},
+		})
+	}()
+
+	// Broadcast the new block to the network.
+	if err := bs.Broadcaster.Broadcast(ctx, v1alpha1Block); err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not broadcast block: %v", err)
+	}
+
+	if err := bs.BlockReceiver.ReceiveBlock(ctx, v1alpha1Block, root); err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not process beacon block: %v", err)
+	}
+
+	return &ptypes.Empty{}, nil
 }
 
 // GetBlock retrieves block details for given block id.
 func (bs *Server) GetBlock(ctx context.Context, req *ethpb.BlockRequest) (*ethpb.BlockResponse, error) {
-	return nil, errors.New("unimplemented")
+	blk, err := bs.blockFromBlockID(ctx, req.BlockId)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get block from block ID: %v", err)
+	}
+	if blk == nil {
+		return nil, status.Errorf(codes.NotFound, "Could not find requested block")
+	}
+
+	v1Block, err := migration.V1Alpha1ToV1Block(blk)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not convert block to v1")
+	}
+
+	return &ethpb.BlockResponse{
+		Data: &ethpb.BeaconBlockContainer{
+			Message:   v1Block.Block,
+			Signature: blk.Signature,
+		},
+	}, nil
 }
 
 // GetBlockRoot retrieves hashTreeRoot of BeaconBlock/BeaconBlockHeader.
 func (bs *Server) GetBlockRoot(ctx context.Context, req *ethpb.BlockRequest) (*ethpb.BlockRootResponse, error) {
-	return nil, errors.New("unimplemented")
+	var root []byte
+	var err error
+	switch string(req.BlockId) {
+	case "head":
+		root, err = bs.HeadFetcher.HeadRoot(ctx)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Could not retrieve head block: %v", err)
+		}
+		if root == nil {
+			return nil, status.Errorf(codes.NotFound, "No head root was found")
+		}
+	case "finalized":
+		finalized := bs.FinalizationFetcher.FinalizedCheckpt()
+		if finalized == nil {
+			return nil, status.Errorf(codes.NotFound, "No finalized root was found")
+		}
+		root = finalized.Root
+	case "genesis":
+		blk, err := bs.BeaconDB.GenesisBlock(ctx)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Could not retrieve blocks for genesis slot: %v", err)
+		}
+		if blk == nil {
+			return nil, status.Error(codes.NotFound, "Could not find genesis block")
+		}
+		blkRoot, err := blk.Block.HashTreeRoot()
+		if err != nil {
+			return nil, status.Error(codes.Internal, "Could not hash genesis block")
+		}
+		root = blkRoot[:]
+	default:
+		if len(req.BlockId) == 32 {
+			block, err := bs.BeaconDB.Block(ctx, bytesutil.ToBytes32(req.BlockId))
+			if err != nil {
+				return nil, status.Errorf(codes.Internal, "Could not retrieve block for block root %#x: %v", req.BlockId, err)
+			}
+			if block == nil {
+				return nil, status.Error(codes.NotFound, "Could not find any blocks with given root")
+			}
+
+			root = req.BlockId
+		} else {
+			slotInt, err := strconv.Atoi(string(req.BlockId))
```

Comment:
```suggestion
			slotInt, err := strconv.ParseUint(string(req.BlockId), 10, 64)
```
Then you don't have to cast to unit64